### ISSUE TITLE
[Bug] ModelListGP and SingleTaskGP have different hyperparameters

### ISSUE THREAD

**Issue by MarkusTMaier** [pytorch/botorch#1150]
# 🐛 Bug

Hi all,

I am modelling several different outputs using the SingleTaskGP model. I wanted to convert the SingleTaskGP to a ModelListGP because I would like to exclude some inputs from the model (the idea was to do this by specifying active_dims of the kernel). However, I noticed that by switching from the SingleTaskGP to the ModelListGP the hyperparameters of the model and consequently my optimization significantly changed. The lengthscale and noise hyperparameters did not change much but the output scale hyperparameters are around 0.65 for the ModelListGP and around 0.86 for the SingleTask GP. I would expect that the hyperparameters should be similar for the SingleTaskGP and the ModelListGP. Thanks for your help.

## To reproduce

```python
from gpytorch.kernels.scale_kernel import ScaleKernel
from gpytorch.priors.torch_priors import GammaPrior
from botorch.models import SingleTaskGP, ModelListGP
from gpytorch.kernels.matern_kernel import MaternKernel
from gpytorch.likelihoods.gaussian_likelihood import GaussianLikelihood
from gpytorch.mlls.sum_marginal_log_likelihood import SumMarginalLogLikelihood
from gpytorch.constraints.constraints import GreaterThan
from gpytorch import ExactMarginalLogLikelihood
from botorch import fit_gpytorch_model
from botorch.models.transforms import Normalize
import torch


def get_covar_module(ard_num_dims, batch_shape):
    covar_module1 = ScaleKernel(
        MaternKernel(
            nu=2.5,
            ard_num_dims=ard_num_dims,
            batch_shape=batch_shape,
            lengthscale_prior=GammaPrior(10.0, 30.0)
        ),
        batch_shape=batch_shape,
        outputscale_prior=GammaPrior(10.0, 10.0)
    )
    covar_module2 = ScaleKernel(
        MaternKernel(
            nu=2.5,
            ard_num_dims=ard_num_dims,
            batch_shape=batch_shape,
            lengthscale_prior=GammaPrior(3.0, 6.0)
        ),
        batch_shape=batch_shape,
        outputscale_prior=GammaPrior(10.0, 10.0)
    )
    covar_module = covar_module1 + covar_module2
    return covar_module

def get_likelihood(batch_shape):
    noise_prior = GammaPrior(1, 5)
    noise_prior_mode = (noise_prior.concentration - 1) / noise_prior.rate
    likelihood = GaussianLikelihood(
        noise_prior=noise_prior,
        batch_shape=batch_shape,
        noise_constraint=GreaterThan(
            1e-4,
            transform=None,
            initial_value=noise_prior_mode,
        ),
    )
    return likelihood

train_input = torch.tensor([[4.5262e+00, 7.1968e+00, 4.8251e+00, 2.7234e+00, 4.5219e+00, 2.9192e+00,
         6.1045e-01, 1.4911e+00, 4.6071e-01, 8.9810e-01, 6.8157e+01, 1.6704e+00],
        [2.0794e+00, 3.5391e+00, 5.8643e+00, 1.9830e+01, 3.4140e+00, 5.7206e-01,
         2.9001e+00, 1.0809e+00, 2.2235e-01, 5.9277e-01, 5.8181e+01, 1.7240e+00],
        [6.5846e-01, 3.0880e+00, 5.5571e+00, 2.6360e+01, 1.1747e-01, 5.6796e-01,
         3.5031e+00, 9.4447e-01, 1.9026e-01, 5.5981e-01, 5.6654e+01, 1.7995e+00],
        [3.7033e-01, 1.4448e+01, 1.2863e+00, 4.8949e-01, 2.8790e+00, 6.6686e+00,
         3.7033e+00, 8.7498e-01, 2.0968e-02, 5.2146e-01, 6.5246e+01, 3.4915e+00],
        [3.9385e+00, 9.5642e-01, 1.4120e+01, 6.4151e+00, 4.9533e+00, 4.2480e+00,
         2.6074e+00, 1.1607e+00, 5.8308e-01, 1.1134e-01, 5.8376e+01, 2.5306e+00],
        [3.5226e+00, 1.7139e+01, 2.3917e+00, 1.3678e+00, 1.7678e+00, 3.3306e+00,
         5.5148e+00, 1.2408e+00, 7.2490e-01, 9.4666e-03, 5.9720e+01, 3.2706e+00],
        [5.6120e-01, 1.2980e+01, 4.9009e+00, 3.4253e+00, 2.7694e+00, 8.1942e+00,
         2.6736e+00, 9.3776e-01, 3.6869e-01, 3.4813e-01, 5.9763e+01, 3.0779e+00],
        [5.8859e+00, 1.3754e+00, 7.2704e-01, 7.7399e-01, 6.8049e+00, 4.2493e+00,
         6.6335e+00, 1.2624e+00, 3.0446e-01, 9.1203e-01, 6.7432e+01, 3.6390e+00],
        [5.1180e-02, 7.1087e+00, 1.5893e+01, 1.2722e+00, 1.4359e+00, 8.3224e-01,
         4.6377e+00, 7.9401e-01, 2.3084e-01, 5.8370e-01, 6.4810e+01, 2.3504e+00],
        [2.6607e+00, 8.2045e+00, 7.1150e-01, 5.8007e+00, 3.5391e+00, 3.6948e+00,
         1.1536e+01, 6.1346e-01, 2.4430e-01, 2.5515e-01, 5.9572e+01, 3.1678e+00]])

train_outcome = torch.tensor([[0.4448, 0.6845, 0.1654, 0.9065, 1.0068, 0.6909, 0.3299, 0.3342],
        [0.3769, 0.8715, 0.1759, 0.9245, 1.0242, 0.5913, 0.3975, 0.3585],
        [0.4212, 0.6815, 0.2638, 0.8058, 0.9947, 0.6224, 0.3354, 0.3921],
        [0.4140, 0.6228, 0.0658, 0.9013, 0.8430, 0.5670, 0.2611, 0.3672],
        [0.3890, 0.6978, 0.1749, 0.9949, 0.9150, 0.5987, 0.3317, 0.3568],
        [0.3805, 0.7427, 0.2511, 1.0057, 0.9487, 0.3802, 0.2952, 0.5119],
        [0.2593, 0.7673, 0.1598, 1.0335, 0.8953, 0.7678, 0.3485, 0.3709],
        [0.3883, 0.6824, 0.1671, 0.9852, 1.0031, 0.7249, 0.2642, 0.3748],
        [0.4387, 0.7334, 0.0328, 1.0402, 0.9105, 0.6522, 0.2947, 0.3852],
        [0.3979, 0.6494, 0.2432, 1.0722, 0.9327, 0.6357, 0.2770, 0.3648]])

bounds = torch.tensor([[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.5, 0.0, 0.0, 50.0, 1.5],
                               [10.0, 50.0, 50.0, 50.0, 50.0, 50.0, 50.0, 1.5, 1.5, 1.0, 85.0, 4.0]],
                              dtype=torch.double)


batch_shape = torch.Size([train_outcome.shape[1]])
model = SingleTaskGP(train_X=train_input,
                     train_Y=train_outcome,
                     likelihood=get_likelihood(batch_shape=batch_shape),
                     covar_module=get_covar_module(ard_num_dims=train_input.shape[1], batch_shape=batch_shape),
                     input_transform=Normalize(d=train_input.shape[1], bounds=bounds)
                     )
model = model.to(dtype=torch.double)
model_likelihood = ExactMarginalLogLikelihood(model.likelihood, model)
fit_gpytorch_model(model_likelihood, sequential=False, options={"disp": True})

print('Hyperparameters batch model:')
print('Length scales kernel 0:')
print(model.covar_module.kernels[0].base_kernel.lengthscale)
print('Length scales kernel 1:')
print(model.covar_module.kernels[1].base_kernel.lengthscale)
print('Outputscale kernel 0:')
print(model.covar_module.kernels[0].outputscale)
print('Outputscale kernel 1:')
print(model.covar_module.kernels[1].outputscale)
print('Noise:')
print(model.likelihood.noise)


model_list = []
for nx in range(train_outcome.shape[1]):
    model = SingleTaskGP(train_X=train_input,
                         train_Y=train_outcome[:, nx].unsqueeze(1),
                         likelihood=get_likelihood(batch_shape=torch.Size([])),
                         covar_module=get_covar_module(ard_num_dims=train_input.shape[1], batch_shape=torch.Size([])),
                         input_transform=Normalize(d=train_input.shape[1], bounds=bounds))
    model_list.append(model)
model = ModelListGP(*model_list)
model_likelihood = SumMarginalLogLikelihood(model.likelihood, model)
fit_gpytorch_model(model_likelihood, sequential=False, options={"disp": True})
print('Hyperparameters model list:')
print('Length scales kernel 0:')
print(model.models[0].covar_module.kernels[0].base_kernel.lengthscale)
print(model.models[1].covar_module.kernels[0].base_kernel.lengthscale)
print('Length scales kernel 1:')
print(model.models[0].covar_module.kernels[1].base_kernel.lengthscale)
print(model.models[1].covar_module.kernels[1].base_kernel.lengthscale)
print('Outputscale kernel 0:')
print(model.models[0].covar_module.kernels[0].outputscale)
print(model.models[1].covar_module.kernels[0].outputscale)
print('Outputscale kernel 1:')
print(model.models[0].covar_module.kernels[1].outputscale)
print(model.models[1].covar_module.kernels[1].outputscale)
print('Noise:')
print(model.models[0].likelihood.noise)
print(model.models[1].likelihood.noise)
```

## Expected Behavior

I would expect that the hyperparameters of the SingleTaskGP and the ModelListGP are very similar.

## System information

**Please complete the following information:**
- 0.5.1
- 1.6.0
- 1.10.0
-  Windows Server 2019 Standard

**Comment by qingfeng10** [comment]
Hey @MarkusTMaier, thanks for reporting this!  Looking at your example, I think there are couple reasons that you saw different output-scales from SingleTaskGP and the ModelListGP. 
1. The hyper-parameter fitting (`fit_gpytorch_model`) might end up with slight different values. There is no guarantee to get exact same values. In particular, when the loss function is hard to optimize and it is easily hitting local optimum. 

2. In your case, the model fitting looks hard in terms of the parameter dimension (~12 parameters) and there is only 10 training samples.  It looks like you construct an additive kernel. This gives even more hyper-parameters to fit and it is not quite identifiable.  I guess, what happens during fitting is returning any local optimum. I would recommend to try different kernels (with better sample efficiency).  
> the idea was to do this by specifying active_dims of the kernel 
It seems like you do want to turn off some parameters for fitting. But that does not reflect in the code you share. Here is an example if you want to specify the active parameters (via active_dims) 

```
covar_module = ScaleKernel(
    base_kernel=MaternKernel(
    nu=2.5,
    ard_num_dims=3,
    active_dims=torch.tensor([0, 1, 2]),
         lengthscale_prior=GammaPrior(3.0, 6.0),
    ),
    outputscale_prior=GammaPrior(2.0, 0.5),
)
```

3. The SingleTaskGP here has a relatively strong priors on the Kernel hyperparameters, which work best when covariates are normalized to the unit cube and outcomes are standardized (zero mean, unit variance). Therefore, standardization of outcomes can really help. 

```
from botorch.models.transforms.outcome import Standardize

model = SingleTaskGP(
    train_X=train_input,
    train_Y=train_outcome,
    likelihood=get_likelihood(batch_shape=batch_shape),
    covar_module=get_covar_module(ard_num_dims=train_input.shape[1], batch_shape=batch_shape),
    input_transform=Normalize(d=train_input.shape[1], bounds=bounds),
    outcome_transform=Standardize(m=8),
)
```

Considering all these, the difference is more for a fitting issue rather than a bug. I made a comparison of hyperparameters values using default covar_module in SingleTaskGP and including the outcome_transform.  Below is what I got 

```
from botorch.models.transforms.outcome import Standardize

model = SingleTaskGP(train_X=train_input,
    train_Y=train_outcome,
    # likelihood=get_likelihood(batch_shape=batch_shape),
    # covar_module=get_covar_module(ard_num_dims=train_input.shape[1], batch_shape=batch_shape),
    input_transform=Normalize(d=train_input.shape[1], bounds=bounds),
    outcome_transform=Standardize(m=train_outcome.shape[1]),
)
mll = ExactMarginalLogLikelihood(model.likelihood, model)
fit_gpytorch_model(mll);

print(model.covar_module.outputscale)
# tensor([0.9193, 0.6673, 1.0825, 0.9899, 0.8961, 1.1357, 0.9678, 1.1542], grad_fn=<SoftplusBackward0>)


model_list = []
for nx in range(train_outcome.shape[1]):
    m = SingleTaskGP(
        train_X=train_input,
        train_Y=train_outcome[:, nx].unsqueeze(1),
        #  likelihood=get_likelihood(batch_shape=torch.Size([])),
        #  covar_module=get_covar_module(ard_num_dims=train_input.shape[1], batch_shape=torch.Size([])),
        input_transform=Normalize(d=train_input.shape[1], bounds=bounds),
        outcome_transform=Standardize(m=1),
    )
    model_list.append(m.to(dtype=torch.double))
model2 = ModelListGP(*model_list)
mll = SumMarginalLogLikelihood(model2.likelihood, model2)
fit_gpytorch_model(mll);

for i in range(8):
    print(model2.models[i].covar_module.outputscale)
# tensor(0.9193, dtype=torch.float64, grad_fn=<SoftplusBackward0>)
# tensor(0.6673, dtype=torch.float64, grad_fn=<SoftplusBackward0>)
# tensor(1.0825, dtype=torch.float64, grad_fn=<SoftplusBackward0>)
# tensor(0.9899, dtype=torch.float64, grad_fn=<SoftplusBackward0>)
# tensor(0.8961, dtype=torch.float64, grad_fn=<SoftplusBackward0>)
# tensor(1.1357, dtype=torch.float64, grad_fn=<SoftplusBackward0>)
# tensor(0.9678, dtype=torch.float64, grad_fn=<SoftplusBackward0>)
# tensor(1.1541, dtype=torch.float64, grad_fn=<SoftplusBackward0>)
```

@MarkusTMaier let me know whether this makes sense or whether you have more questions.

**Comment by MarkusTMaier** [comment]
@qingfeng10 thanks for your quick and detailed answer. I am aware of the points you mentioned. However, for my specific problem I really would like to use a custom covar_module. In the meantime, I did some further examinations. I think the issue is with fit_gpytorch_model(). The example you showed in the end is the standard case. fit_gpytorch_model() converts the batch model (model with multiple outputs) to a model list. Afterwards, the hyperparameters are determined by MAP estimation for each model in the list sequentially. If you directly specify a model list no conversion is necessary (your second example). Also, for this case each model in the list is fitted sequentially. This makes sense for me and is confirmed by your experiments. However, if you have a batch model and at the same time specify a custom covar_module things get tricky. In this case, fit_gpytorch_model() tries to convert the batch model to a model list. In my example, the model conversion fails, probably due to the custom covar_module. Due to this failure fit_gpytorch_model() sets the parameter sequential to False. Now we are not determining the hyperparameters sequentially. I do not fully understand how this works. Is it similar to fitting a multi-task model with independent outputs? On the other hand, if you specify a ModelList with a custom covar_module, fit_gpytorch_model() does not need to convert the model and determines the hyperparameters sequentially. I think the different results of the hyperparameters are caused by the two different fitting approaches. However, I am still surprised by the big difference. I do not think that the difference is caused by a local optimum because I sampled the initial points from the hyperpriors (not shown in code) and this issue is repeatable (it seems that fit_gpytorch_model() also samples the initial points from the hyperpriors). I really think it has something do with the two different fitting procedures. For any hint on this issue I would be grateful.

**Comment by Balandat** [comment]
> Now we are not determining the hyperparameters sequentially. I do not fully understand how this works. Is it similar to fitting a multi-task model with independent outputs? 

In a sense, yes, but for a basic MTGP with an ICM kernel you usually share a single data kernel, whereas in this setting you have multiple kernels (represented as batches). So you'll have to solve a much higher-dimensional optimization problem.

> On the other hand, if you specify a ModelList with a custom covar_module, fit_gpytorch_model() does not need to convert the model and determines the hyperparameters sequentially. I think the different results of the hyperparameters are caused by the two different fitting approaches. However, I am still surprised by the big difference. I do not think that the difference is caused by a local optimum because I sampled the initial points from the hyperpriors (not shown in code) and this issue is repeatable (it seems that fit_gpytorch_model() also samples the initial points from the hyperpriors). 

The difference between fitting the models sequentially (`k` `n`-dim problems) and fitting a batched model (a single `kn`-dim problem) can indeed be substantial, we've observed this in our own work. This is really the reason why we automatically convert a batched model to a `ModelList` internally in `fit_gpytorch_model` in the first place. 

The difficulty that you are running into here is that it's not easy to do this batched-list-batched conversion automatically for custom kernels / covar modules. One option would be to endow the `Kernel` base module with some conversion methods that comply with a common API but allow the user to implement this functionality for their custom kernels so that this can be used automatically in the fitting tools. This however would require changes to the core GPyTorch classes (or some wrapper on the botorch side that defines this API).

**Comment by MarkusTMaier** [comment]
@Balandat Thanks for explaining and clarifying the difference between fitting models sequentially and fitting batch models. For me it is fine, I was mainly surprised by the big difference.

**Comment by Balandat** [comment]
Yeah, I guess the moral of the story is (as usual) that if you know of a particular structure in your optimization problem, exploit it...
